### PR TITLE
Refactor BalancedAllocator.Balancer to LocalShardsBalancer

### PR DESCRIPTION
**Pull request by kotwanikunal**
Signed-off-by: Kunal Kotwani <kkotwani@amazon.com>

### Description
- Refactors `BalancedShardsAllocator.Balancer` to a new class `LocalShardsBalancer`
- Adds a new abstract class called `ShardsBalancer` in prep for `RemoteShard` support
- More details: #4759

### Issues Resolved
- #4776

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java [hunk 1]
```diff
@@ -0,0 +1,967 @@
...
+
+import static org.opensearch.cluster.routing.ShardRoutingState.RELOCATING;
+
+/**
+ * A {@link LocalShardsBalancer} used by the {@link BalancedShardsAllocator} to perform allocation operations
+ * for local shards within the cluster.
+ *
+ * @opensearch.internal
+ */
+public class LocalShardsBalancer extends ShardsBalancer {
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:51]: I'm not sure how conservative we need to be here, but in the interest of compatibility it seems like we could leave the inner class 'Balancer' in place with its current name and just change it to extend the newly introduced interface. These things are tagged `opensearch.internal` but that doesn't mean they aren't used somewhere (though I couldn't find any usage of Balancer in any other repo in opensearch-project). What do you think?

Comment (Bukhtawar) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:51]: 1. Why do we want to expose the `Balancer` outside `BalancedShardAllocator`
2. Can we use a factory for `Balancer` creation?

Comment (andrross) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:51]: > Why do we want to expose the Balancer outside BalancedShardAllocator

I don't think we want to, but it is already used in `org.opensearch.cluster.routing.allocation.AllocationConstraints` which requires it to be public.

> Can we use a factory for Balancer creation?

The approach here (#4759) calls for having 2 balancers with one running after the other, so I'm not sure a factory would help.

Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:51]: > I'm not sure how conservative we need to be here, but in the interest of compatibility it seems like we could leave the inner class 'Balancer' in place with its current name and just change it to extend the newly introduced interface. These things are tagged `opensearch.internal` but that doesn't mean they aren't used somewhere (though I couldn't find any usage of Balancer in any other repo in opensearch-project). What do you think?

I wanted to improve the readability and maintainability by splitting it across appropriate classes. We can keep `LocalShardsBalancer` and `BalancedShardsAllocator.Balancer` and deprecate the latter. We can get rid of it with 3.0.0.

**Review by andrross — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java (hunk 1 above)
Comment (mch2) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:51]: nit - Make this class final?

**Review by mch2 — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java (hunk 1 above)
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java:51]: That's a good suggestion. I recently updated the PR to add back the `Balancer` class to maintain BWC, which essentially extends `LocalShardsBalancer`.
I'll do that with v3.0 when we get rid of the `Balancer` class.

**Pull request merged**